doc/design.rst

```diff
@@ -0,0 +1,113 @@
+*********************
+
+We edit text on the (say) left of the screen, in vim, and review output on the
+(say) right of the screen, in some version of the IPython notebook.
+
+This is *edit left, look right*.
+
+We stored the results that are the basis for the display on the right, as an
+IPython notebook, in notebook JSON format.  We use this notebook as store for
+the results, and we will fetch the results from this store when creating the
+built version of the RST page.
+
+We will call the stored JSON notebook ``scratch.ipynb``, although its actual
+name will probably be some version of the input RST document.
+
+Math / markdown rendering
+=========================
+
+This can either be:
+
+* The current selected block of text; or
+* (No text selected) all text above and below the cursor up to (down to) any
+  directive apart from a ``..math`` directive, or a heading.
+
+We convert this text according to markdown according to the following rules:
+
+* We discard any directives or inline text roles that are not ``..math::`` or
+  ``:math:``;
+* We replace inline or directive ``math`` with text between dollars or text
+  between double dollar markers respectively;
+
+The resulting cell appears last in the rendered notebook.
+
+Code execution and rendering
+============================
+
+Code for execution is in custom code directives::
+
+    .. icode::
+
+        a = 10
+        print(a)
+
+On running some vim function such as ``:ExecICode``, vim does the following:
+
+* finds the code surrounding the cursor position and stores as
+  ``code-block``;
+* creates a UUID for this block if not already specified in the directive
+  options;
+* writes the UUID into the text of the directive, to form something like::
+
+    .. icode::
+        :uuid: 12ab23cd45ef
+
+        a = 10
+        print(a)
+
+* executes this code in the ipython notebook to the right.  The execution code
+  first checks the notebook to see if there is a cell with metadata containing
+  this UUID value.  If so, the code input and output text go into the
+  pre-existing cell in the notebook.  Otherwise the notebook creates and fills a
+  new cell.  The UUID appears in the metadata for the cell in the
+  ``scratch.ipynb`` notebook;
+* I (the user) can execute a command like ``:AlwaysShow`` in the text buffer,
+  such that the result of running this cell, whether refreshing and old cell or
+  creating a new one continues to be shown on the right in the web browswer,
+  even when other cells get executed or rendered.  ``scratch.ipynb`` stores the
+  ``always-show`` flag with the metadata for the results of the executed cell.
+  As above, the association between RST directive instances and
+  ``scratch.ipynb`` cells is via the directive UUID.
+* I may want to play with the code in cell in the rendered ``scratch.ipynb``
+  notebook.  I can do this, then switch back to the vim buffer, and execute some
+  command such as ``:PullCell`` from an ``.. icode`` directive that will find
```

Comment:
ooh, hadn't considered a PullCell - seems like a really good idea! also a generic PullActiveCell would allow grabbing code and output from the left and bring them to the right - making it possible to combine portions of several notebooks into one (something very difficult right now, not even [NB-CCP](https://github.com/ivanov/nb-cccp) does that yet)